### DISCUSSION TITLE
What guarantees do on_connect and on_disconnect make of eachother?

### DISCUSSION THREAD

**Discussion by Keheck** [Rapptz/discord.py#9645] [Q&A]
I wanna know if it is guaranteed that an on_connect event is called only when the bot is started or if there has been an on_disconnect event recently (about a few minutes earlier)
I need to know this because my serialization and deserialization method depend on the behavior of on_connect amd on_disconnect, and having an on_connect event fired randomly/unpredictably throughout the bot's lifecycle would cause issues with my serializable data

**Comment by No767** [comment]
@Keheck Generally you **should never depend** on `on_connect`, and `on_disconnect`. These do get called multiple times randomly ([as per docs](https://discordpy.readthedocs.io/en/stable/api.html#discord.on_ready)). I would be interested to understand what data are you serializing and deserialization during those events. 

P.S: You can join the [official discord.py server](https://discord.gg/r3sSKJJ) if you want quicker response times. Make sure to make a help post within the server

**Reply by Keheck** [reply]
I use these events to de/serialize stuff like reaction roles and custom server stuff that members can get like a custom role. I could probably serialize them in the proper events and deserialize when the bot gets set up now that I think about it

**Comment by Rapptz** [comment]
There's no real guarantee on these events. `on_connect` is called whenever a connection is attempted and `on_disconnect` is called whenever a disconnect happens. Sometimes you can disconnect multiple times before you finally connect so it's not a 1:1 mapping.
